compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -603,12 +603,6 @@ fn layout_of_uncached<'tcx>(
                     .flatten()
             };
 
-            let dont_niche_optimize_enum = def.repr().inhibit_enum_layout_opt()
-                || def
-                    .variants()
-                    .iter_enumerated()
-                    .any(|(i, v)| v.discr != ty::VariantDiscr::Relative(i.as_u32()));
```

Comment:
We lost the following comment during refactorings in the past:

> The current code for niche-filling relies on variant indices instead of actual discriminants, so enums with explicit discriminants (RFC #2363) would misbehave.